src/mpi/comm/comm_impl.c

```diff
@@ -514,6 +514,21 @@ static int get_tag_from_stringtag(const char *stringtag)
     return hash % (MPIR_Process.attrs.tag_ub);
 }
 
+#ifndef MPID_SESSION_USE_WORLD
+static int calc_context_id(MPIR_Group * group_ptr, const char *stringtag)
```

Comment:
This seems fine for now, but could we make PMI the authority for bootstrapping context ids somehow in the future?

Reply:
I'm only suggesting for the bootstrap stage. Having a static `MPIR_CTXID_BOOTSTRAP` proved not to be sufficient. I was wondering if we could allocate a bootstrap ID from PMI and use it in place of this function (which we could still use as a fallback). Not critical now but we could add it as a TODO.

Reply:
From offline discussion, if the groups in question are managed by the PMI server then a PMI assigned ctxid could work. But since users are allowed to construct groups independently from PMI, there really isn't a good way to do the allocation without another collective PMI call that may overlap with other operations. In short, just rely on the user to use unique string tags and the hash function.